### PR TITLE
Add validation for the task.issue command

### PR DESCRIPTION
**Pull request by DenisNikulin5**
In this PR I added the logic to validate the task.issue command message from the task-lib.
The purpose of the changes is to separate the command executions caused by the task-lib and by using the related logging command in the user script.

To do this, the agent generates a random token every pipeline run, then it will be captured by the task-lib. The task-lib adds the token to every task.issue command message. The token is validated on the agent side in the end.

Related PR: https://github.com/microsoft/azure-pipelines-task-lib/pull/1021

### CODE REVIEW

**Review by KonstantinTyukalov — APPROVED** [review]

File: src/Agent.Worker/TaskRunner.cs [hunk 1]
```diff
@@ -673,5 +684,38 @@ private void PublishTelemetry(Definition taskDefinition, HandlerData handlerData
...
+                    var version_prop = jobj["version"].ToString();
+
+                    // I use the TaskVersion class here just for version comparison.
+                    var version = new TaskVersion(Regex.Match(version_prop, @"\d+\.\d+\.\d+").ToString());
+                    return version >= new TaskVersion() { Major = 4, Minor = 10, Patch = 0 };
+                }
+            }
+            catch(Exception e)
+            {
+                throw new InvalidOperationException($"Coudn't get the task SDK version for task {task.Data.Name} {task.Data.Version.Major}.{task.Data.Version.Minor}.{task.Data.Version.Patch}.", e);
```
Comment (merlynomsft) [inline src/Agent.Worker/TaskRunner.cs:715]: This should be a diagnostics log / telemetry; can't verify on every task

File: src/Microsoft.VisualStudio.Services.Agent/Constants.cs [hunk 2]
```diff
@@ -88,6 +88,7 @@ public static string PathVariable
         public static string PluginTracePrefix = "##[plugin.trace]";
         public static readonly int AgentDownloadRetryMaxAttempts = 3;
         public const string projectName = "projectName";
+        public const string TaskSDKCommandTokenEnvVar = "TASK_SDK_COMMAND_TOKEN";
```
Comment (merlynomsft) [inline src/Microsoft.VisualStudio.Services.Agent/Constants.cs:91]: we need a better name for this.  E.g. CorrelationId or LogId instead of 'token'

File: src/Agent.Worker/TaskRunner.cs [hunk 3]
```diff
@@ -673,5 +681,29 @@ private void PublishTelemetry(Definition taskDefinition, HandlerData handlerData
                 ExecutionContext.Debug(ex.ToString());
             }
         }
+
+        private bool IsTaskSDKTokenRequired(IHandler handler, Definition task)
```
Comment (merlynomsft) [inline src/Agent.Worker/TaskRunner.cs:685]: This is pretty complex, some diagnostics tracing might be helpful if something goes wrong

Comment (DenisNikulin5) [reply src/Agent.Worker/TaskRunner.cs]: Done

File: src/Microsoft.VisualStudio.Services.Agent/Constants.cs (hunk 2 above)
Comment (DenisNikulin5) [reply src/Microsoft.VisualStudio.Services.Agent/Constants.cs]: Changed the name to the COMMAND_CORRELATION_ID

File: src/Agent.Worker/TaskCommandExtension.cs [hunk 4]
```diff
@@ -371,6 +371,21 @@ public void Execute(IExecutionContext context, Command command)
             var eventProperties = command.Properties;
             var data = command.Data;
 
+            bool tokenValidationRequired = false;
```
[the remark this answers is not in the document]
Comment (DenisNikulin5) [reply src/Agent.Worker/TaskCommandExtension.cs]: Done

File: src/Agent.Worker/TaskManager.cs [hunk 5]
```diff
@@ -410,6 +411,68 @@ public sealed class Definition
...
+                return null;
+            }
+
+            var versionLine = File.ReadAllLines(modulePath).FirstOrDefault(x => x.Contains("ModuleVersion"));
+            if (string.IsNullOrEmpty(versionLine))
+            {
+                return null;
+            }
+
+            var verRegex = new Regex(@"\d+\.\d+\.\d+");
```
Comment (merlynomsft) [inline src/Agent.Worker/TaskManager.cs:429]: >verRegex [](http://example.com/codeflow?start=16&length=8)

I'm thinking this goes a bit far to avoid exceptions in TaskVersion.  It's unfortunate that Task Version doesn't have a TryParse, but we can just use catch(ArgumentException) in this case.

At the same time, it also doesn't preview all cases where TaskVersion might throw an exception

File: src/Agent.Worker/TaskManager.cs [hunk 6]
```diff
@@ -410,6 +411,68 @@ public sealed class Definition
...
+            if (!verRegex.IsMatch(versionLine))
+            {
+                return null;
+            }
+
+            var version = new TaskVersion(verRegex.Match(versionLine).Value)
+            {
+                IsTest = new Regex("(?i)(preview|test)").IsMatch(versionLine)
```
Comment (merlynomsft) [inline src/Agent.Worker/TaskManager.cs:437]: >test [](http://example.com/codeflow?start=49&length=4)

pretty sure TaskVersion already handles this?

Comment (merlynomsft) [inline src/Agent.Worker/TaskManager.cs:437]: >IsTest [](http://example.com/codeflow?start=16&length=6)

pretty sure TaskVersion already handles this?

**Review by merlynomsft — APPROVED** [review]
:shipit:

File: src/Agent.Worker/TaskManager.cs (hunk 5 above)
Comment (DenisNikulin5) [reply src/Agent.Worker/TaskManager.cs:429]: I know this logic might seem a bit overloaded. I implemented it this way not to avoid exceptions, but to ensure correct behavior.

Originally, the TaskVersion class was designed to handle versions of agent tasks, not package versions. I decided to repurpose it to avoid adding additional logic for version comparison (especially since it includes a necessary 'IsTest' flag).

So, if I directly input the '5.0.1-preview.0' version into the TaskVersion constructor method, an exception will be raised since the TaskVersion doesn't expect the word 'preview' or more than three delimiter signs ('.' or '-').

In short, this requires preprocessing.

File: src/Agent.Worker/TaskManager.cs (hunk 6 above)
Comment (DenisNikulin5) [reply src/Agent.Worker/TaskManager.cs:437]: That's just a bool flag which can be used for comparison. No special handling needed.

Comment (DenisNikulin5) [reply src/Agent.Worker/TaskManager.cs:437]: The Task Version could set it automatically if the version has suffix '-test' but not for '-peview'. So, again, pre-processing is needed.
